[WIP] Global skaffold config, version 2

nkubala opened the pull request:
This is a second variant on #886. Main differences:

* changes the structure of the config to have a top-level "global" section, followed by any number of context-specific config entries, inspired by @balopat. Example:
```
global: 
  default-repo: gcr.io/myrepo1/
kubecontexts:
   - kubectx: minikube
         default_repo: gcr.io/myrepo2/
   - kubectx: my_dev_gke
         default_repo: gcr.io/myrepo2/
```
Setting global values in the config is done through the `--global` flag, i.e. `skaffold config set default-repo my-global-repo --global`.
* disallows setting config values for multiple contexts at a time
* instead of specifying `--kubecontext all` to show all config values, a `--all` flag is added to the `list` command. `skaffold config list --all`

Opening this PR to show what this implementation would look like.

balopat commented on cmd/skaffold/app/cmd/config_test.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+// func TestSetConfig(t *testing.T) {
```
do you want to remove this block or still hack on it?

nkubala replied:
whoops, nope I wrote this as an integration test so I'll remove it

r2d4 commented on integration/run_test.go:
```diff
@@ -231,3 +238,153 @@ func TestFix(t *testing.T) {
+	}
+}
+
+func randomString() string {
```
how about https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/util/util.go#L33?

nkubala replied:
perfect

r2d4 commented on Gopkg.lock:
```diff
@@ -1160,10 +1152,10 @@
     "github.com/pkg/errors",
     "github.com/sirupsen/logrus",
     "github.com/spf13/cobra",
+    "github.com/spf13/pflag",
```
where is this dependency coming from?

nkubala replied:
this is just an artifact added by the dep solver. the actual pflag dependency didn't change here

r2d4 commented on cmd/skaffold/app/cmd/config/util.go:
```diff
@@ -0,0 +1,115 @@
+	yaml "gopkg.in/yaml.v2"
+)
+
+const defaultConfigLocation = ".skaffold/config"
```
will this work on windows?

nkubala replied:
it should, I see windows-specific logic in `github.com/mitchellh/go-homedir` which is what I'm using here along with `filepath.Join`

r2d4 replied:
Do you need to do a filepath.Join() on this? Should it be `.skaffold\config` on windows?

nkubala replied:
good catch, I should have thought about that when I responded :)

nkubala commented on cmd/skaffold/app/cmd/config/flags.go:
```diff
@@ -0,0 +1,37 @@
+
+func AddConfigFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&configFile, "config", "c", "", "path to skaffold config")
+	cmd.Flags().StringVarP(&kubectx, "kubectx", "k", "", "kubectl context to set values against")
```
yeah I'm fine with that

nkubala commented on cmd/skaffold/app/cmd/config/config.go:
```diff
@@ -0,0 +1,27 @@
+}
+
+type ContextConfig struct {
+	Kubectx     string `yaml:"kubectx,omitempty"`
```
done

nkubala commented on cmd/skaffold/app/cmd/config/util.go:
```diff
@@ -0,0 +1,115 @@
+	}
+	_, err = os.Stat(configFile)
+	// TODO(nkubala): create default config?
+	if err != nil {
```
decided to use `os.OpenFile` here to create a default if it doesn't exist

nkubala commented:
@balopat 
> * if the folder is not created it throws errors in all the subcommands

Folder now gets autocreated along with config file if it doesn't exist
> * no list for possible configs - would be helpful instead of just providing a "this is not a default config"

Most other config management CLIs don't do this (kubectl, gcloud, etc.) since it requires introspecting about the config struct at runtime. I could add this but I think it's probably better to just document it somewhere and use that as a reference.
> * no feedback around which context I set the value for - it would be great to have the feedback when I set the config regarding what I changed as it can be surprising

Updated config value with context is now logged to the user

r2d4 commented on cmd/skaffold/app/cmd/config/set.go:
```diff
@@ -0,0 +1,118 @@
+		RunE: func(cmd *cobra.Command, args []string) error {
+			resolveKubectlContext()
+			err := setConfigValue(args[0], args[1])
+			if err == nil {
```
might be more idiomatic to do

```
if err != nil {
  return err
}
logConfigChangeForUser(out, args[0], args[1])
```

nkubala replied:
I actually had this before and changed it, because it's one less line to do it this way: the way you described you also need to do a `return nil` afterwards. curious to hear your opinion on this

r2d4 replied:
slight preference for the return nil, but up to you

r2d4 commented on cmd/skaffold/app/cmd/config/util.go:
```diff
@@ -0,0 +1,140 @@
+	context, err := context.CurrentContext()
+	if err != nil {
+		logrus.Warn(errors.Wrap(err, "retrieving current kubectl context"))
+		kubecontext = "default"
```
add this as a constant somewhere?

nkubala replied:
actually I reworked this logic to fall back to the global values (not a "default" context) if one isn't set, thanks for pointing this out. updated

r2d4 commented on cmd/skaffold/app/cmd/config/util.go:
```diff
@@ -0,0 +1,140 @@
+func resolveConfigFile() error {
+	if configFile != "" {
+		// we had a config provided as a flag, expand it and return
+		if !filepath.IsAbs(configFile) {
```
why does this need to be an absolute path? add the reasoning as a comment?

nkubala replied:
thanks, this was from some previous logic which I removed a while back so I'll take this out

r2d4 commented on cmd/skaffold/app/cmd/config/config.go:
```diff
@@ -0,0 +1,27 @@
+
+package config
+
+type Config struct {
```
add a docstring here so its clear this is unrelated to SkaffoldConfig

nkubala replied:
done
